# UTS Sistem Temu Kembali Informasi
**Nama:** Suluh Yoga Pratama  
**NIM:** A11.2023.14979  
**Kelas:** A11.4703  
**Topik:** Implementasi Boolean Model dan Vector Space Model

## Import dan Setup

In [13]:
!pip install nltk Sastrawi
import nltk
nltk.download('punkt')
nltk.download('stopwords')
nltk.download('punkt_tab')

[nltk_data] Downloading package punkt to /root/nltk_data...
[nltk_data]   Package punkt is already up-to-date!
[nltk_data] Downloading package stopwords to /root/nltk_data...
[nltk_data]   Package stopwords is already up-to-date!
[nltk_data] Downloading package punkt_tab to /root/nltk_data...
[nltk_data]   Unzipping tokenizers/punkt_tab.zip.


True

In [14]:
import sys, os
sys.path.append('/content/src')

## Preprocessing Dokumen

In [15]:
from src.preprocess import process_folder

# pastikan folder data/raw sudah diupload ke /content/data/raw
process_folder('data/raw', 'data/processed')

[OK] doc39.txt → processed (63 tokens)
[OK] doc18.txt → processed (63 tokens)
[OK] doc36.txt → processed (63 tokens)
[OK] doc10.txt → processed (63 tokens)
[OK] doc1.txt → processed (64 tokens)
[OK] doc32.txt → processed (63 tokens)
[OK] doc45.txt → processed (62 tokens)
[OK] doc15.txt → processed (63 tokens)
[OK] doc14.txt → processed (64 tokens)
[OK] doc9.txt → processed (62 tokens)
[OK] doc38.txt → processed (63 tokens)
[OK] doc4.txt → processed (64 tokens)
[OK] doc20.txt → processed (63 tokens)
[OK] doc5.txt → processed (62 tokens)
[OK] doc43.txt → processed (64 tokens)
[OK] doc47.txt → processed (63 tokens)
[OK] doc24.txt → processed (64 tokens)
[OK] doc44.txt → processed (64 tokens)
[OK] doc23.txt → processed (64 tokens)
[OK] doc29.txt → processed (62 tokens)
[OK] doc6.txt → processed (63 tokens)
[OK] doc31.txt → processed (63 tokens)
[OK] doc11.txt → processed (63 tokens)
[OK] doc41.txt → processed (64 tokens)
[OK] doc34.txt → processed (64 tokens)
[OK] doc16.txt → processed (63

In [16]:
# contoh lihat hasil preprocess
with open('data/processed/doc1.txt', 'r', encoding='utf-8') as f:
    print(f.read()[:300])

konteks sistem temu informasi sistem informasi modern butuh mekanisme akses temu dokumen relevan efisien dekat tradisional andal cocok kunci sederhana metode tangkap makna konseptual dekat baru manfaat model representasi teks kaya teknik bobot tfidf algoritma perankingan hasil cari urut dasar muncul


## Boolean Retrieval Model

In [17]:
from src.boolean_ir import load_processed_docs, build_inverted_index, boolean_search

docs = load_processed_docs('data/processed')
index = build_inverted_index(docs)
all_docs = list(docs.keys())

# uji beberapa query
print("Query 1:", boolean_search("informasi AND sistem", index, all_docs))
print("Query 2:", boolean_search("NOT evaluasi", index, all_docs))
print("Query 3:", boolean_search("(indeks AND vektor) OR evaluasi", index, all_docs))

Query 1: {'doc35.txt', 'doc21.txt', 'doc36.txt', 'doc17.txt', 'doc42.txt', 'doc25.txt', 'doc7.txt', 'doc16.txt', 'doc28.txt', 'doc1.txt', 'doc11.txt', 'doc26.txt', 'doc13.txt', 'doc14.txt', 'doc6.txt', 'doc5.txt', 'doc50.txt', 'doc40.txt', 'doc48.txt', 'doc37.txt', 'doc46.txt', 'doc15.txt', 'doc47.txt', 'doc31.txt', 'doc41.txt', 'doc44.txt', 'doc24.txt', 'doc9.txt', 'doc27.txt', 'doc22.txt', 'doc18.txt', 'doc49.txt', 'doc34.txt', 'doc2.txt', 'doc38.txt', 'doc20.txt', 'doc4.txt', 'doc43.txt', 'doc10.txt', 'doc29.txt', 'doc39.txt', 'doc12.txt', 'doc32.txt', 'doc45.txt', 'doc30.txt', 'doc3.txt', 'doc23.txt', 'doc33.txt', 'doc8.txt', 'doc19.txt'}
Query 2: {'doc35.txt', 'doc21.txt', 'doc36.txt', 'doc17.txt', 'doc42.txt', 'doc25.txt', 'doc7.txt', 'doc16.txt', 'doc28.txt', 'doc1.txt', 'doc11.txt', 'doc26.txt', 'doc13.txt', 'doc14.txt', 'doc6.txt', 'doc5.txt', 'doc50.txt', 'doc40.txt', 'doc48.txt', 'doc37.txt', 'doc46.txt', 'doc15.txt', 'doc47.txt', 'doc31.txt', 'doc41.txt', 'doc44.txt', 'doc2

## Vector Space Model (TF-IDF)

In [18]:
from src.vsm_ir import compute_tfidf, search_vsm

tfidf_docs, idf = compute_tfidf(docs)
hasil = search_vsm("informasi sistem dokumen", tfidf_docs, idf)
hasil[:10]

[('doc1.txt', 0.42600643361512924),
 ('doc41.txt', 0.42600643361512924),
 ('doc21.txt', 0.42600643361512924),
 ('doc20.txt', 0.38023455031468684),
 ('doc40.txt', 0.38023455031468684),
 ('doc18.txt', 0.3559840029391736),
 ('doc38.txt', 0.3559840029391736),
 ('doc45.txt', 0.3227486121839514),
 ('doc5.txt', 0.3227486121839514),
 ('doc25.txt', 0.3227486121839514)]

## Evaluasi (Precision, Recall, MAP, nDCG)

In [20]:
from src.eval import precision_recall, precision_at_k, average_precision, ndcg_at_k

gold = ['doc1.txt', 'doc2.txt', 'doc3.txt']
res = [h[0] for h in hasil[:5]]

p, r, tp, fp, fn = precision_recall(res, gold)
print("Precision:", p, "Recall:", r)
print("P@5:", precision_at_k(hasil, gold, k=5))
print("AP:", average_precision(hasil, gold))
print("nDCG:", ndcg_at_k(hasil, gold))

Precision: 0.2 Recall: 0.3333333333333333
P@5: 0.2
AP: 0.3333333333333333
nDCG: 0.46927872602275644


## Penjelasan Hasil Evaluasi

Berdasarkan hasil evaluasi menggunakan metrik Information Retrieval, sistem memperoleh Precision sebesar 0.20 dan Recall sebesar 0.33. Nilai-nilai tersebut menunjukkan bahwa sistem masih cenderung menampilkan dokumen yang tidak relevan serta belum mampu menemukan seluruh dokumen relevan yang ada. Nilai AP (Average Precision) sebesar 0.33 dan nDCG (Normalized Discounted Cumulative Gain) sebesar 0.47 mengindikasikan bahwa urutan hasil pencarian masih perlu ditingkatkan agar dokumen relevan dapat muncul pada peringkat lebih tinggi.